targets/simple_switch/externs/sym_crypto/crypto.cpp

```diff
@@ -0,0 +1,618 @@
+    // hexdump((char*)&integrity_check_value[0], integrity_check_value.size());
+
+    std::vector<unsigned char> user_data;
+    user_data.reserve(secure_data_size);
+
+    int valid = validation_function(secure_association_key,
+                                    secure_channel_identifier,
+                                    packet_number,
+                                    destination_mac_address,
+                                    source_mac_address,
+                                    security_tag, secure_data,
+                                    integrity_check_value, user_data);
+
+
+    // cout << "[p4sec] user_data" << std::endl;
+    // hexdump((char*)&user_data[0], user_data.size());
+
+    // cout << "[p4sec] Ethertype" << std::endl;
+    // hexdump((char*)&user_data[0], ETHERTYPE_SIZE);
+
+    // cout << "[p4sec] decrypted payload" << std::endl;
+    // hexdump((char*)&user_data[ETHERTYPE_SIZE],
+    // user_data.size() - ETHERTYPE_SIZE);
+
+    // replace payload
+    // first, remove all the data
+    get_packet().remove(get_packet().get_data_size());
+    // make room for the ciphertext and write the ciphertext in it
+    char *payload_start = get_packet().prepend(
+                          static_cast<uint64_t> (user_data.size())
+                          - ETHERTYPE_SIZE);
+    for (uint i = 0; i < user_data.size() - ETHERTYPE_SIZE; i++) {
+      payload_start[i] = user_data[i + ETHERTYPE_SIZE];
+    }
+
+    // copy ethertype from encrypted packet
+    std::stringstream ss_ethertype;
+    for (uint i = 0; i < ETHERTYPE_SIZE; ++i)
+        ss_ethertype << std::setfill('0') << std::setw(2)
+        << std::hex << static_cast<int>(user_data[i]);
+    std::string ethertype_hexstr = ss_ethertype.str();
+
+    out_ethertype.set(ethertype_hexstr);
+    out_valid.set(valid);
+  }
+
+  std::vector<unsigned char> get_char_vector(string str, uint size) {
+    // string fitted_str = fit_string(str, size);
+    std::vector<unsigned char> vec(size, '\0');
+    if (str.length() > size) {
+      // cout << "[p4sec] given string was too long" << std::endl;
+      str.resize(size);
+    }
+    vec.insert(vec.cend()-size, str.begin(), str.end());
+
+    return vec;
+  }
+
+  void protection_function(std::vector<unsigned char> secure_association_key,
+                           std::vector<unsigned char> secure_channel_identifier,
+                           std::vector<unsigned char> packet_number,
+                           std::vector<unsigned char> destination_mac_address,
+                           std::vector<unsigned char> source_mac_address,
+                           std::vector<unsigned char> security_tag,
+                           std::vector<unsigned char> user_data,
+                           std::vector<unsigned char>& out_secure_data,
+                           std::vector<unsigned char>& out_integrity_check_value
+                          ) {
+    // hier evtl assertions fuer die Laenge der Parameter
```

Comment:
this comment is in German?

Reply:
Yes, the authors are at a university in Germany.  I can change the comment to English.